src/Hooks/GutenbergCleanup.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace WPML\PB\Elementor\Hooks;
+
+use WPML\FP\Fns;
+use WPML\FP\Maybe;
+use WPML\FP\Obj;
+use WPML\FP\Relation;
+use WPML\PB\Elementor\DataConvert;
+use WPML_Elementor_Data_Settings;
+use function WPML\FP\curryN;
+use function WPML\FP\partial;
+use function WPML\FP\pipe;
+
+class GutenbergCleanup implements \IWPML_Backend_Action, \IWPML_Frontend_Action, \IWPML_DIC_Action {
+
+	public function add_hooks() {
+		add_filter(
+			'update_post_metadata',
+			Fns::withoutRecursion( Fns::identity(), [ $this, 'removeGutenbergFootprint' ] ),
+			10,
+			4
+		);
+	}
+
+	/**
+	 * If we detect Gutenberg footprint in the Elementor data,
+	 * we'll remove it and delete the Gutenberg string package.
+	 *
+	 * @param null|bool $check
+	 * @param int       $postId
+	 * @param string    $metaKey
+	 * @param string    $metaValue
+	 *
+	 * @return mixed
+	 */
+	public function removeGutenbergFootprint( $check, $postId, $metaKey, $metaValue ) {
+		if (
+			WPML_Elementor_Data_Settings::META_KEY_DATA === $metaKey
+			&& WPML_Elementor_Data_Settings::is_edited_with_elementor( $postId )
+		) {
+			$data = DataConvert::unserialize( $metaValue );
+
+			// $ifDifferentFrom :: array -> array -> bool
+			$ifDifferentFrom = curryN( 2, function( $data,  $processedData )  {
+				return json_encode( $processedData ) !== json_encode( $data );
+			} );
+
+			// $updatePostMeta :: array -> array
+			$updatePostMeta = Fns::tap( pipe(
+				[ DataConvert::class, 'serialize' ],
+				partial( 'update_metadata', 'post', $postId, WPML_Elementor_Data_Settings::META_KEY_DATA )
+			) );
+
+			// $deleteGbPackage :: array -> array
+			$deleteGbPackage = Fns::tap( pipe(
+				Fns::always( $postId ),
+				[ self::class, 'getGutenbergPackage' ],
+				[ self::class, 'deletePackage' ]
+			) );
+
+			return Maybe::of( $data )
+				->map( [ $this, 'removeBlockMetaInEditorWidget' ] )
+				->filter( $ifDifferentFrom( $data ) )
+				->map( $updatePostMeta )
+				->map( $deleteGbPackage )
+				->map( Fns::always( true ) )
+				->getOrElse( $check );
```

Comment:
@strategio Something doesn't seem right with this. I understand why you have done it this way. I would try to make it something like:

```php
return Maybe::of( $data )
				->map( [ $this, 'removeBlockMetaInEditorWidget' ] )
				->filter( $ifDifferentFrom( $data ) )
				->map( [ DataConvert::class, 'serialize' ] )
				->map( $updatePostMeta( $postId ) )
				->map( $deleteGbPackage )
				->getOrElse( $check );
```
and change $updatePostMeta to return $postId and $deleteGbPackage to return true.

This way the data flows through the functions.

Also $updatePostMeta and $deleteGbPackage are then pure functions and don't take data from the surrounding environment.

Reply:
Please check again @brucepearson.

Reply:
@strategio I was thinking more like this. This way the data is passed from function to function and each function returns transformed data for the next function.

```php
			// $updatePostMeta :: int -> array -> int
			$updatePostMeta = curryN( 2, function( $postId, $meta ) {
				Post::updateMeta( $postId, WPML_Elementor_Data_Settings::META_KEY_DATA, $meta );
				return $postId;
			} );

			return Maybe::of( $data )
				->map( [ $this, 'removeBlockMetaInEditorWidget' ] )
				->filter( $ifDifferentFrom( $data ) )
				->map( [ DataConvert::class, 'serialize' ] )
				->map( $updatePostMeta( $postId ) )
				->map( [ self::class, 'getGutenbergPackage' ] )
				->map( [ self::class, 'deletePackage' ] )
				->getOrElse( $check );

...
...

public static function deletePackage( $package ) {
	$package && do_action( 'wpml_delete_package', $package->name, $package->kind );
	return true;    
}
```

The other option would be to have all the `side effects` in one function:

```php
			// $update :: int -> array -> bool
			$update = curryN( 2, function( $postId, $meta ) {
				Post::updateMeta( $postId, WPML_Elementor_Data_Settings::META_KEY_DATA, $meta );
				self::deletePackage( self::getGutenbergPackage( $postId ) );
				return true;
			} );

			return Maybe::of( $data )
				->map( [ $this, 'removeBlockMetaInEditorWidget' ] )
				->filter( $ifDifferentFrom( $data ) )
				->map( [ DataConvert::class, 'serialize' ] )
				->map( $update( $postId ) )
				->getOrElse( $check );

```

Reply:
@brucepearson, I opted for the 2nd solution. Please check again!

Reply:
@strategio Looks good now. Don't forget to delete commented out code!

Reply:
Ooops, good catch @brucepearson, thanks :)